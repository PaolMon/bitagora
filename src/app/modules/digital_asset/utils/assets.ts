import { BaseModuleDataAccess, StateStore } from "lisk-sdk";
import { codec } from "lisk-sdk";
import { digitalAsset, registeredAssets } from "../../../schemas/digital_asset/digital_asset_types";
import { registeredAssetsSchema } from "../../../schemas/digital_asset/digital_asset_schemas";

export const CHAIN_STATE_DIGITAL_ASSETS = "digitalAsset:registeredAssets";


export const setNewAsset= async (stateStore: StateStore, asset: digitalAsset) => {

    const DAs:digitalAsset[] = await getAllAssets(stateStore);

    DAs.push(asset);

    const registeredAssets: registeredAssets = {registeredAssets: DAs.sort((a, b) => a.merkleRoot.compare(b.merkleRoot))};

    await stateStore.chain.set(
        CHAIN_STATE_DIGITAL_ASSETS,
        codec.encode(registeredAssetsSchema, registeredAssets)
    );
}

export const getAllAssets = async (stateStore: StateStore) => {
    const registeredAssetsBuffer = await stateStore.chain.get(
        CHAIN_STATE_DIGITAL_ASSETS
    );

    if (!registeredAssetsBuffer) {
        return [];
    }

    const registeredAssets = codec.decode<registeredAssets>(
        registeredAssetsSchema,
        registeredAssetsBuffer
    );

    return registeredAssets.registeredAssets;
}

export const getAssetByMerkleRoot = async (stateStore: StateStore, merkleRoot: Buffer) => {
    const DAs:digitalAsset[] = await getAllAssets(stateStore);

    const DA_index: number = DAs.findIndex((t) => t.merkleRoot.equals(merkleRoot));

    if (DA_index < 0) {
        throw new Error("Asset not found");
    }

    return DAs[DA_index];
}

/*
    THE FOLLOWING FUNCTIONS USE DATA ACCESS INSTEAD OF STATE STORE TO READ DATA
*/


export const _getAssetByMerkleRoot = async (dataAccess: BaseModuleDataAccess, merkleRoot: Buffer): Promise<digitalAsset> => {
    const registeredAssetsBuffer = await dataAccess.getChainState(
        CHAIN_STATE_DIGITAL_ASSETS
    );
    
    if (!registeredAssetsBuffer) {
        throw new Error('No Assets Found in stateStore');
    }
    
    const registeredAssets = codec.decode<registeredAssets>(
        registeredAssetsSchema,
        registeredAssetsBuffer
    );
    const DAs = registeredAssets.registeredAssets;

    const DA_index: number = DAs.findIndex((t) => t.merkleRoot.equals(merkleRoot));

    if (DA_index < 0) {
        throw new Error("Asset not found");
    }

    return DAs[DA_index];
}

export const _getAllJSONAssets =async (dataAccess: BaseModuleDataAccess) => {
    const registeredAssetsBuffer = await dataAccess.getChainState(
        CHAIN_STATE_DIGITAL_ASSETS
    );
    
    if (!registeredAssetsBuffer) {
        return [];
    }
    
    const registeredAssets = codec.decode<registeredAssets>(
        registeredAssetsSchema,
        registeredAssetsBuffer
    );
    
      return codec.toJSON(registeredAssetsSchema, registeredAssets);
}

export const _getAssetHistoryByMerkleRoot = async (dataAccess: BaseModuleDataAccess, merkleRoot: Buffer) => {

    const registeredAssetsBuffer = await dataAccess.getChainState(
        CHAIN_STATE_DIGITAL_ASSETS
    );
    
    if (!registeredAssetsBuffer) {
        return [];
    }
    
    const registeredAssets = codec.decode<registeredAssets>(
        registeredAssetsSchema,
        registeredAssetsBuffer
    );
    const DAs = registeredAssets.registeredAssets;

    let related_assets: digitalAsset[] = []
    let index = -1;
    let asset: digitalAsset;

    while(!merkleRoot.equals(Buffer.alloc(0))) {
        
        index = DAs.findIndex((t) => t.merkleRoot.equals(merkleRoot));

        if (index < 0) {
            throw new Error("Asset not found");
        }

        asset = DAs[index];
        related_assets.push(asset);

        merkleRoot = asset.previousAssetReference;
    }

    return codec.toJSON(registeredAssetsSchema, {registeredAssets: related_assets});
    //return related_assets;
}