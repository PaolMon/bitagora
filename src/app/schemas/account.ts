export interface BitagoraAccountProps {
    infos: {
        myNeighbors: {
            address: Buffer,
            ipAddress: string,
            port: number
        }[],
        hostedFiles: {
            owner: Buffer,
            merkleRoot: Buffer
        }[],
    },
    digitalAsset: {
        pending: {
            fileName: string,
            merkleRoot: Buffer
        }[],
        allowed: {
            fileName: string,
            merkleRoot: Buffer,
            secret: string
        }[],
        myFiles: {
            fileName: string,
            merkleRoot: Buffer,
            secret:string
        }[],
    }
}

export const infosSchema = {
    type: 'object',
    properties: {
        myNeighbors: {
            fieldNumber: 1,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    address: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    ipAddress: {
                        dataType:'string',
                        fieldNumber: 2
                    },
                    port: {
                        dataType:'uint32',
                        fieldNumber: 3
                    }
                }
            }
        },
        hostedFiles: {
            fieldNumber: 2,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    owner: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    merkleRoot: {
                        dataType:'bytes',
                        fieldNumber: 2
                    }
                }
            }
        }
    },
    default: {
        myNeighbors: [],
        hostedFiles: []
    }
}

export const digitalAssetAccountSchema = {
    type: 'object',
    properties: {
        myFiles: {
            fieldNumber: 1,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    merkleRoot: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    fileName: {
                        dataType:'string',
                        fieldNumber: 2
                    },
                    secret: {
                        dataType:'string',
                        fieldNumber: 3
                    }
                }
            }
        },
        pending: {
            fieldNumber: 2,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    merkleRoot: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    fileName: {
                        dataType:'string',
                        fieldNumber: 2
                    }
                }
            }
        },
        allowed: {
            fieldNumber: 3,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    merkleRoot: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    fileName: {
                        dataType:'string',
                        fieldNumber: 2
                    },
                    secret: {
                        dataType:'string',
                        fieldNumber: 3
                    }
                }
            }
        }
    },
    default: {
        myFiles: [],
        pending: [],
        allowed: []
    }
}
