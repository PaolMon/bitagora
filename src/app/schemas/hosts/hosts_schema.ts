export const registeredHostsSchema = {
	$id: 'lisk/infos/registeredHosts',
	type: "object",
	required: ["hosts"],
	properties: {
        hosts: {
		type: "array",
		fieldNumber: 1,
		items: {
			type: 'object',
			required: ['address', 'ipAddress', 'port'],
			properties: {
				address: {
					dataType: 'bytes',
					fieldNumber: 1
				},
				ipAddress:{
					dataType: 'string',
					fieldNumber: 2
				},
				port: {
					dataType: 'uint32',
					fieldNumber: 3
				}
			},
		}
	  }
	}
	
};