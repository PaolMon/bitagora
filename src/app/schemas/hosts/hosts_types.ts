
export type registered_hosts = {
    hosts: host[]
}

export type host = {
    address: Buffer,
    ipAddress: string,
    port: number
}