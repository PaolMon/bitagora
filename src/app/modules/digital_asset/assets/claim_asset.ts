import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';
import { digitalAsset } from '../../../schemas/digital_asset/digital_asset_types';
import { getAssetByMerkleRoot, setNewAsset } from '../utils/assets';
import { chunk } from '../../../schemas/chunks/chunk_types';
import { getChunkByMerkleRoot, setNewChunk } from '../utils/chunks';
import { BitagoraAccountProps } from '../../../schemas/account';


export type claim = {
	oldMerkleRoot: Buffer,
	newMerkleRoot: Buffer,
	newMerkleHeight: number,
	newHosts: Buffer[],
	newSecret: string
}

export class ClaimAsset extends BaseAsset {
	public name = 'claim';
  	public id = 3;

  	// Define schema for asset
	public schema = {
		$id: 'digitalAsset/claim-asset',
		title: 'ClaimAsset transaction asset for digitalAsset module',
		type: 'object',
		required: ['oldMerkleRoot','newMerkleRoot', 'newMerkleHeight', 'newHosts'], 
		properties: {
			oldMerkleRoot:{
				dataType: 'bytes',
				fieldNumber: 1
			},
			newMerkleRoot:{
				dataType: 'bytes',
				fieldNumber: 2
			},
			newMerkleHeight: {
				dataType: 'uint32',
				fieldNumber: 3
			},
			newHosts: {
				type: 'array',
				fieldNumber: 4,
				items: {
					dataType: 'bytes'
				}
			},
			newSecret: {
				dataType: 'string',
				fieldNumber: 5
			}
		},
	};

	public validate({ asset }: ValidateAssetContext<claim>): void {
		// Validate your asset
		if (asset.newMerkleHeight === 0 && asset.newHosts.length < 3) {
			throw new Error("If file has been splitted into chunks, it must be sent to at least 3 hosts");
		}
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<claim>): Promise<void> {
		
		const senderAddress: Buffer = transaction.senderAddress;
		const old_chunk: chunk = await getChunkByMerkleRoot(stateStore, asset.oldMerkleRoot);

		if (old_chunk.owner.compare(senderAddress) != 0) {
			throw new Error("You are not allowed to claim this asset");
		}

		const digital_asset: digitalAsset = await getAssetByMerkleRoot(stateStore, asset.oldMerkleRoot);

		digital_asset.owner = senderAddress;
		digital_asset.merkleRoot = asset.newMerkleRoot;
		digital_asset.merkleHeight = asset.newMerkleHeight;
		digital_asset.previousAssetReference = asset.oldMerkleRoot;
		digital_asset.transactionID = transaction.id;
		digital_asset.secret = asset.newSecret;
	
		const chunk: chunk = {
			owner: senderAddress,
			merkleRoot: asset.newMerkleRoot,
			hostedBy: [],
			requestedBy: [],
			allowedViewers: []
		} 
		
		const senderAccount = await stateStore.account.get<BitagoraAccountProps>(senderAddress);
		
		senderAccount.digitalAsset.myFiles.push({fileName: digital_asset.fileName, merkleRoot: digital_asset.merkleRoot, secret: asset.newSecret})

		await setNewChunk(stateStore, chunk);

		await setNewAsset(stateStore, digital_asset);

		await stateStore.account.set(senderAccount.address, senderAccount).catch();

	}
}
