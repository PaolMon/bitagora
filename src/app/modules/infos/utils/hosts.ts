import {  codec, StateStore } from "lisk-sdk";
import { registeredHostsSchema } from "../../../schemas/hosts/hosts_schema";
import { host, registered_hosts } from "../../../schemas/hosts/hosts_types";

export const CHAIN_STATE_HOSTS_ADDR = "infos:registeredHosts";

export const getHostIndex =async (list:host[], address: Buffer): Promise<number> => {

    const index: number = list.findIndex((t) => t.address.equals(address));

    if (index < 0) {
        throw new Error("Host not found");
    }

    return index;
}

export const setNewHost= async (stateStore: StateStore, host: host) => {

    const list:host[] = await getAllHosts(stateStore);

    list.push(host);

    const registeredHosts: registered_hosts = {hosts: list.sort((a, b) => a.address.compare(b.address))};

    await stateStore.chain.set(
        CHAIN_STATE_HOSTS_ADDR,
        codec.encode(registeredHostsSchema, registeredHosts)
    );
}

export const getAllHosts = async (stateStore: StateStore) => {
    const registeredHostsBuffer = await stateStore.chain.get(
        CHAIN_STATE_HOSTS_ADDR
    );

    if (!registeredHostsBuffer) {
        return [];
    }

    const registeredHosts = codec.decode<registered_hosts>(
        registeredHostsSchema,
        registeredHostsBuffer
    );

    return registeredHosts.hosts;
}