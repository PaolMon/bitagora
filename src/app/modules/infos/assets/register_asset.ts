import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';
import { host } from '../../../schemas/hosts/hosts_types';
import { getAllHosts, getHostIndex, setNewHost } from '../utils/hosts';

import { BitagoraAccountProps } from '../../../schemas/account';

export type register = {
	ipAddress: string,
	port: number
}

const MAX_NEIGHBORS = 4;

export class RegisterAsset extends BaseAsset {
	public name = 'register';
  	public id = 0;

  // Define schema for asset
	public schema = {
    	$id: 'infos/register-asset',
		title: 'RegisterAsset transaction asset for infos module',
		type: 'object',
		required: ['ipAddress', 'port'],
		properties: {
			ipAddress: {
				dataType: 'string',
				fieldNumber: 1
			},
			port: {
				dataType: 'uint32',
				fieldNumber: 2
			}
		},
  	};

	public validate({ asset }: ValidateAssetContext<register>): void {
		const splitted: string[] = asset.ipAddress.split('.');
		if(splitted.length !== 4) {
			throw new Error('Not a valid IP Address.');
		}
		for (const octet of splitted) {
			if (Number(octet).toString() === 'NaN' || Number(octet).toString(2).length > 8) {
				throw new Error('Not a valid IP Address.');
			}
		}
		if (asset.port > 65535) { //da verificare numero massimo di porta
			throw new Error('Not a valid Port number.');
		}
	}

	// eslint-disable-next-line @typescript-eslint/require-await
  	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<register>): Promise<void> {
		const senderAddress:Buffer = transaction.senderAddress;

		await setNewHost(stateStore, {
			address: senderAddress,
			ipAddress: asset.ipAddress,
			port: asset.port
		});


		const list: host[] = await getAllHosts(stateStore);

		const h_index = await getHostIndex(list, senderAddress)
		
		let begin = h_index - MAX_NEIGHBORS/2;
		let end = h_index + MAX_NEIGHBORS/2

		if( begin < 0) {
			begin = 0
			end = MAX_NEIGHBORS + 1
		}
		if( end >= list.length) {
			begin = list.length - 1 - MAX_NEIGHBORS
			end = list.length - 1
		}

		let my_neighbors: host[] = [...list.slice(begin,h_index), ...list.slice(h_index+1, end+1)];

		let senderAccount = await stateStore.account.get<BitagoraAccountProps>(senderAddress);

		senderAccount.infos.myNeighbors = my_neighbors;

		await stateStore.account.set(senderAddress, senderAccount).catch();
	}
}
