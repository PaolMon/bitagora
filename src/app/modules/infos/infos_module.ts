/*
 * LiskHQ/lisk-commander
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

/* eslint-disable class-methods-use-this */

import {
    AfterBlockApplyContext,


    AfterGenesisBlockApplyContext, BaseModule,


    BeforeBlockApplyContext, StateStore, TransactionApplyContext
} from 'lisk-sdk';
import { BitagoraAccountProps, infosSchema } from '../../schemas/account';
import { RegisterAsset } from "./assets/register_asset";

export class InfosModule extends BaseModule {
    public actions = {
        // Example below
        // getBalance: async (params) => this._dataAccess.account.get(params.address).token.balance,
        // getBlockByID: async (params) => this._dataAccess.blocks.get(params.id),
        getAccountInfos:async (params:Record<string, unknown>) => {
            let { address } = params;
            if (!Buffer.isBuffer(address) && typeof address === 'string') {
                address = Buffer.from(address, 'hex')
            } 
            const account = await this._dataAccess.getAccountByAddress<BitagoraAccountProps>(address as Buffer);
            return account.infos;
        },
        getAccount:async (params:Record<string, unknown>) => {
            let { address } = params;
            if (!Buffer.isBuffer(address) && typeof address === 'string') {
                address = Buffer.from(address, 'hex')
            } 
            const account = await this._dataAccess.getAccountByAddress<BitagoraAccountProps>(address as Buffer);
            return JSON.stringify(account);
        },
    };
    public reducers = {
        // Example below
        hostChunk: async (
			params: Record<string, unknown>,
			stateStore: StateStore,
		): Promise<void> => {
			const { senderAddress, owner, merkleRoot } = params;
			if (!Buffer.isBuffer(senderAddress)) {
				throw new Error('senderAddress must be a buffer');
			}
			if (!Buffer.isBuffer(owner)) {
				throw new Error('owner must be a buffer');
			}
			if (!Buffer.isBuffer(merkleRoot)) {
				throw new Error('merkle_root must be a buffer');
			}
			const account = await stateStore.account.get<BitagoraAccountProps>(senderAddress);
            account.infos.hostedFiles.push({
                owner: owner,
                merkleRoot: merkleRoot
            })

            await stateStore.account.set(senderAddress, account)
		},
    };
    public name = 'infos';
    public accountSchema = infosSchema;
    public transactionAssets = [new RegisterAsset()];
    public events = [
        // Example below
        // 'infos:newBlock',
        'newHost',
    ];
    public id = 1002;

    // public constructor(genesisConfig: GenesisConfig) {
    //     super(genesisConfig);
    // }

    // Lifecycle hooks
    public async beforeBlockApply(_input: BeforeBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async afterBlockApply(_input: AfterBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async beforeTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
        if (_input.transaction.moduleID === this.id && _input.transaction.assetID === 0) {
            this._channel.publish('infos:newHost', {address: _input.transaction.senderAddress})
        }
    }

    public async afterGenesisBlockApply(_input: AfterGenesisBlockApplyContext) {
        // Get any data from genesis block, for example get all genesis accounts
        // const genesisAccoounts = genesisBlock.header.asset.accounts;
    }
}
