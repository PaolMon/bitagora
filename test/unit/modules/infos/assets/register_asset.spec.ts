import { ApplyAssetContext, codec, StateStore, testing } from 'lisk-sdk';
import { Account } from '@liskhq/lisk-chain';
import { register, RegisterAsset } from '../../../../../src/app/modules/infos/assets/register_asset';
import { InfosModule } from '../../../../../src/app/modules/infos/infos_module';
import { BitagoraAccountProps } from '../../../../../src/app/schemas/account';
import { registered_hosts } from '../../../../../src/app/schemas/hosts/hosts_types';
import { CHAIN_STATE_HOSTS_ADDR } from '../../../../../src/app/modules/infos/utils/hosts';
import { registeredHostsSchema } from '../../../../../src/app/schemas/hosts/hosts_schema';

describe('RegisterAsset', () => {
  let transactionAsset: RegisterAsset;

	beforeEach(() => {
		transactionAsset = new RegisterAsset();
	});

	describe('constructor', () => {
		it('should have valid id', () => {
			expect(transactionAsset.id).toEqual(0);
		});

		it('should have valid name', () => {
			expect(transactionAsset.name).toEqual('register');
		});

		it('should have valid schema', () => {
			expect(transactionAsset.schema).toMatchSnapshot();
		});
	});

	describe('validate', () => {
		describe('schema validation', () => {
			it.todo('should throw errors for invalid schema');
			it.todo('should be ok for valid schema');
		});
	});

	describe('apply', () => {
		let stateStore: StateStore;
		let account: Account<BitagoraAccountProps>;
		let context: ApplyAssetContext<register>;
		let test_registered_hosts: registered_hosts;
		
		beforeEach(() => {
			account = testing.fixtures.createDefaultAccount<BitagoraAccountProps>([InfosModule]);
			
			test_registered_hosts = {
				hosts: [
					{
						address: Buffer.alloc(9,'address_1'),
						ipAddress: '192.168.1.10',
						port: 9999
					},
					{
						address: Buffer.alloc(9,'address_2'),
						ipAddress: '192.168.1.20',
						port: 8888
					},
					{
						address: Buffer.alloc(9,'address_3'),
						ipAddress: '192.168.1.30',
						port: 7777
					},
					{
						address: Buffer.alloc(9,'address_4'),
						ipAddress: '192.168.1.40',
						port: 6666
					},
					// {
					// 	address: Buffer.alloc(9,'address_5'),
					// 	ipAddress: '192.168.1.50',
					// 	port: 5555
					// }
				]
			}
			stateStore = new testing.mocks.StateStoreMock({
				accounts: [account],
				chain: {
					"infos:registeredHosts": codec.encode(registeredHostsSchema, test_registered_hosts),
				}
			});
			
			const register_object: register = {
				ipAddress: '127.0.0.1',
				port: 10000
			}
	
			context = testing.createApplyAssetContext<register>({
				stateStore,
				asset: register_object,
				transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
			});
	
			jest.spyOn(stateStore.chain, 'get');
			jest.spyOn(stateStore.chain, 'set');
		});
		describe('valid cases', () => {
			it('should update state store',async () => {
				await transactionAsset.apply(context);
				let list = test_registered_hosts.hosts;
				list.push({
					address: context.transaction.senderAddress,
					ipAddress: '127.0.0.1',
					port: 10000
				});
				const registeredHosts: registered_hosts = {hosts: list.sort((a, b) => a.address.compare(b.address))};
				expect(stateStore.chain.set).toHaveBeenCalledWith(
					CHAIN_STATE_HOSTS_ADDR, 
					codec.encode(registeredHostsSchema, registeredHosts)
				)
			});
			it('should update account infos',async () => {
				await transactionAsset.apply(context);
				const senderAccount = await stateStore.account.get<BitagoraAccountProps>(context.transaction.senderAddress)
				expect(senderAccount.infos.myNeighbors).toEqual(test_registered_hosts.hosts)
			});
		});

		describe('invalid cases', () => {
			it.todo('should throw error');
		});
	});
});
