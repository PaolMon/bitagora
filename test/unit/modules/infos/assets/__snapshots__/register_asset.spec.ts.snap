// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterAsset constructor should have valid schema 1`] = `
Object {
  "$id": "infos/register-asset",
  "properties": Object {
    "ipAddress": Object {
      "dataType": "string",
      "fieldNumber": 1,
    },
    "port": Object {
      "dataType": "unit32",
      "fieldNumber": 2,
    },
  },
  "required": Array [
    "ipAddress",
    "port",
  ],
  "title": "RegisterAsset transaction asset for infos module",
  "type": "object",
}
`;
